### PR TITLE
metal: Q1_0 backend

### PR DESCRIPTION
**Pull request by khosravipasha**
## Overview

Follow up after merging of [Q1_0 CPU PR](https://github.com/ggml-org/llama.cpp/pull/21273). This PR adds the relevant Metal backend.

These are to speed up familly of Bonsai 1-bit models on the Mac:
- [prism-ml/Bonsai-8B-gguf](https://huggingface.co/prism-ml/Bonsai-8B-gguf)
- [prism-ml/Bonsai-4B-gguf](https://huggingface.co/prism-ml/Bonsai-4B-gguf)
- [prism-ml/Bonsai-1.7B-gguf](https://huggingface.co/prism-ml/Bonsai-1.7B-gguf)
- See a working demo from our Bonsai-demo repo: [Bonsai-demo](https://github.com/prismml-eng/Bonsai-demo)
- See more details here: [1-bit-bonsai-8b-whitepaper.pdf](https://github.com/PrismML-Eng/Bonsai-demo/blob/main/1-bit-bonsai-8b-whitepaper.pdf)

Device: Apple M4 Pro, macOS, Metal backend

### Bonsai-1.7B (231.13 MiB, 1.72B params)

| model                             |       size |     params | backend  | threads | fa |    test |               t/s |
| --------------------------------- | ---------: | ---------: | -------- | ------: | -: | ------: | -----------------: |
| qwen3 1.7B                        | 231.13 MiB |     1.72 B | MTL,BLAS |      10 |  1 |   pp512 | 2306.82 ± 19.64    |
| qwen3 1.7B                        | 231.13 MiB |     1.72 B | MTL,BLAS |      10 |  1 |   tg128 |   307.23 ±  9.99   |

### Bonsai-4B (540.09 MiB, 4.02B params)

| model                             |       size |     params | backend  | threads | fa |    test |               t/s |
| --------------------------------- | ---------: | ---------: | -------- | ------: | -: | ------: | -----------------: |
| qwen3 4B                          | 540.09 MiB |     4.02 B | MTL,BLAS |      10 |  1 |   pp512 |  922.93 ±  2.60    |
| qwen3 4B                          | 540.09 MiB |     4.02 B | MTL,BLAS |      10 |  1 |   tg128 |  189.38 ±  0.38    |

### Bonsai-8B (1.07 GiB, 8.19B params)

| model                             |       size |     params | backend  | threads | fa |    test |               t/s |
| --------------------------------- | ---------: | ---------: | -------- | ------: | -: | ------: | -----------------: |
| qwen3 8B                          |   1.07 GiB |     8.19 B | MTL,BLAS |      10 |  1 |   pp512 |  498.19 ±  3.55    |
| qwen3 8B                          |   1.07 GiB |     8.19 B | MTL,BLAS |      10 |  1 |   tg128 |  124.51 ±  0.67    |

## End-to-end testing: KL Divergence (Q1_0 vs unpacked into FP16)

To test accuray of the Metal backend, we can compare the KL divergence of the Q1_0 model against the unpacked FP16 model. 
The weights are equivalant so checking the logits give us a good indication of the accuracy of the Metal backend (at least for prompt processing). Ran on 200 chunks of wikitext-2-raw, ctx 512. As expected its close to 0, the difference is due to numerical precision probably.

For each model testing vs the unpacked version here: https://huggingface.co/collections/prism-ml/bonsai-auxiliary

| Model | Mean KLD | RMS Δp | Same top p | PPL correlation | pp512 (t/s) | tg128 (t/s) |
|-------|----------|--------|------------|-----------------|-------------|-------------|
| 8B    | 0.000000 | 0.023% | 100.000%   | 99.93%          | 498         | 125         |
| 4B    | 0.000001 | 0.017% | 100.000%   | 99.91%          | 923         | 189         |
| 1.7B  | 0.000001 | 0.018% | 99.922%    | 99.90%          | 2307        | 307         |

# Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: Yes, I agree. AI was mainly used to run validation tests and explore ideas to tune the kernels. I tried the changes manually myself and ran it through our validation tests.

### CODE REVIEW

**Comment by ggerganov** [comment]
> In our original release the llama.cpp Metal Kernel were a bit slower than the MLX version, when submitting this PR had the time to look into why and optimize the llama.cpp Metal a bit more (by 40%-45% almost).

Yes, I had a mental note to ask you to adjust the `N_R0_Q1_0` and `N_SG_Q1_0` constants. I assume this is what brought the most benefit?

File: ggml/src/ggml-metal/ggml-metal.metal [hunk 1]
```diff
@@ -3337,6 +3433,87 @@ void mul_vec_q_n_f32_impl(
...
+    const short ix = (tiisg/8);
+    const short il = (tiisg%8)*16;
+
+    device const float * yb = y + ix*QK1_0 + il;
+
+    for (int ib = ix; ib < nb; ib += N_SIMDWIDTH/8) {
+        float sumy = 0.f;
+
+#pragma unroll
+        for (short i = 0; i < 16; i++) {
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal.metal]:
```suggestion
        FOR_UNROLL (short i = 0; i < 16; i++) {
```

File: ggml/src/ggml-metal/ggml-metal.metal [hunk 2]
```diff
@@ -3337,6 +3433,87 @@ void mul_vec_q_n_f32_impl(
...
+            yl[i] = yb[i];
+            sumy += yb[i];
+        }
+
+#pragma unroll
+        for (short row = 0; row < nr0; row++) {
```
Comment (ggerganov) [inline ggml/src/ggml-metal/ggml-metal.metal]:
```suggestion
        FOR_UNROLL (short row = 0; row < nr0; row++) {
```

**Review by ggerganov — APPROVED** [review]

**Comment by khosravipasha** [comment]
A lot of the speed up also came from just rebasing to new llama.cpp (our branch was 1-2 months old)


Stage | Description | 8B t/s | 4B t/s | 1.7B t/s
-- | -- | -- | -- | --
0 | Original prism branch (older llama.cpp) | 81 | 124 | 232
1 | Rebase to latest llama.cpp (same kernel) | 95 | 143 | 251
2 | select() + unrolled byte access + sumy | 105 | 153 | 269
3 | N_R0=8 tuning | 118 | 178 | 293
